lgc/state/PalMetadata.cpp

```diff
@@ -237,19 +295,102 @@ void PalMetadata::setUserDataEntry(ShaderStage stage, unsigned userDataIndex, un
 // @param dwordOffset : Dword offset that the spill table is used at
 void PalMetadata::setUserDataSpillUsage(unsigned dwordOffset) {
   if (dwordOffset < m_spillThreshold->getUInt())
-    *m_spillThreshold = m_document->getNode(dwordOffset);
+    *m_spillThreshold = dwordOffset;
+}
+
+// =====================================================================================================================
+// Fix up user data registers. Any user data register that has one of the unlinked UserDataMapping values defined
+// in AbiUnlinked.h is fixed up by looking at pipeline state.
+void PalMetadata::fixUpRegisters() {
+  static const std::pair<unsigned, unsigned> ComputeRegRanges[] = {{mmCOMPUTE_USER_DATA_0, 16}};
+  static const std::pair<unsigned, unsigned> Gfx8RegRanges[] = {
+      {mmSPI_SHADER_USER_DATA_PS_0, 16}, {mmSPI_SHADER_USER_DATA_VS_0, 16}, {mmSPI_SHADER_USER_DATA_GS_0, 16},
+      {mmSPI_SHADER_USER_DATA_ES_0, 16}, {mmSPI_SHADER_USER_DATA_HS_0, 16}, {mmSPI_SHADER_USER_DATA_LS_0, 16}};
+  static const std::pair<unsigned, unsigned> Gfx9RegRanges[] = {{mmSPI_SHADER_USER_DATA_PS_0, 32},
+                                                                {mmSPI_SHADER_USER_DATA_VS_0, 32},
+                                                                {mmSPI_SHADER_USER_DATA_ES_0, 32},
+                                                                {mmSPI_SHADER_USER_DATA_HS_0, 32}};
+  static const std::pair<unsigned, unsigned> Gfx10RegRanges[] = {{mmSPI_SHADER_USER_DATA_PS_0, 32},
+                                                                 {mmSPI_SHADER_USER_DATA_VS_0, 32},
+                                                                 {mmSPI_SHADER_USER_DATA_GS_0, 32},
+                                                                 {mmSPI_SHADER_USER_DATA_HS_0, 32}};
+
+  ArrayRef<std::pair<unsigned, unsigned>> regRanges = ComputeRegRanges;
+  if (m_pipelineState->isGraphics()) {
+    if (m_pipelineState->getTargetInfo().getGfxIpVersion().major < 9)
+      regRanges = Gfx8RegRanges;
+    else if (m_pipelineState->getTargetInfo().getGfxIpVersion().major == 9)
+      regRanges = Gfx9RegRanges;
+    else
+      regRanges = Gfx10RegRanges;
+  }
+
+  // First find the descriptor sets and push const nodes.
+  SmallVector<const ResourceNode *, 4> descSetNodes;
+  const ResourceNode *pushConstNode;
+  for (const auto &node : m_pipelineState->getUserDataNodes()) {
+    if (node.type == ResourceNodeType::DescriptorTableVaPtr && !node.innerTable.empty()) {
```

Comment:
This seems to assume that descriptor sets are not in the top level.  However, xgl creates the internal pipelines with descriptor sets in the top level.  See https://github.com/GPUOpen-Drivers/xgl/blob/dev/icd/api/vk_device.cpp#L1931.

Is there any plan to handle this?  This is causes all applications to fail when using relocatable shaders because this compute shader is created for every application.

Reply:
Hmm, good point, and now you mention it, I think I recall that Rex mentioned it too.

Can we avoid using shader compilation on those internal pipelines? Presumably they are amenable to pipeline caching.

Alternatively, and I think this is a question to Qun and Rex, can we stop putting descriptor sets in the top level in internal shaders?

Reply:
Also, we should make ElfLinker::link() spot the situation and return a recoverable error, and enhance llpcCompiler to do something with a recoverable error (chuck away the shader elf(s) and do a pipeline compile).

Reply:
Currently there is no way for the driver to specify which shaders should be compiled using relocatable shaders and which ones should not.  It is a single global option to LLPC.  I've been thinking that we might be able to change that by adding a new option in the pipeline state or something.  We can discuss it later.

We should add something to `canUseRelocatableShader*Elf` to say that this pipeline cannot be compiled using relocatable shaders, and then it will fall back to building the full pipeline.  Anything else we do should be on top of this.

Reply:
Qun will make comments for this.

Reply:
> Hmm, good point, and now you mention it, I think I recall that Rex mentioned it too.
> 
> Can we avoid using shader compilation on those internal pipelines? Presumably they are amenable to pipeline caching.
> 
> Alternatively, and I think this is a question to Qun and Rex, can we stop putting descriptor sets in the top level in internal shaders?

I think it is possible now. until now, we only have one internal compute shader in vk::Device.  please check Device::CreateInternalPipelines (vk_device.cpp) for detail.
but we may encounter more issues for ray-tracing internal pipelines in the future.  please check related internal source code (gpurtInternalShaders.cpp) for detail.

Reply:
I've opened https://github.com/GPUOpen-Drivers/llpc/pull/723 to get this working.

Reply:
FWIW, I'd say having a per-pipeline setting to enable/disable this feature is something we probably want anyway in the long run, to be exposed to applications via the Vulkan API. But that's unrelated to this change.